Fix identity-sdk concurrency issue

koevskinikola opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->

Fixes an issue where too many tenant requeststo Identity cause the service to fail with an `IOException` due to too many concurrent requests.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

closes #23853

koevskinikola committed:
test: reproduce identity-sdk issue

Reproduce an issue where too many tenant requests
to Identity cause the service to fail with an
IOException due to too many concurrent requests.
the gRPC IdentityInterceptor.

The IOException gets lost in the gRPC error translation
but can be seen by debugging the tenant call of the gRPC
IdentityInterceptor.

koevskinikola committed:
feat: implement tenant cache and throttling for identity requests

Implement a tenant cache and throttling for Identity requests
in the gRPC Gateway Interceptor. The cache and throttling are
implemented through a wrapper for the identity-sdk tenants()
method.

The goal is to avoid too many concurrent requests to Identity
that may deteriorate it's service. We are currently aware that
this issue only happens with tenant-related requests. We may
expand this wrapper to other identity-sdk calls, if it is
discovered that too many concurrent calls are an issue there as
well.

This commit only adds a basic implementation to the gRPC API.
Following commits will expand this implementation with config
options. A generalization on the implementation will be applied
so that it can be re-used for the Gateway's REST API server as
well.

koevskinikola committed:
refactor: generalize tenant service implementation

Move the IdentityTenantService class to the gateway module
so that it can be re-used by both the gRPC and REST API
modules.

koevskinikola committed:
feat: add tenant caching and identity throttling to rest

Add tenant caching and throttling of Identity-mananaged
tenant requests. The goal is to reduce the load on Identity
and avoid a high number of concurrent (tenant-related)
requests that have been shown to deteriorate the service.

koevskinikola committed:
feat: add experimental gateway properties

Add experimental Gateway properties to allow
for more unstable features to be used by Zeebe
users. The experimental properties currently
contain configuration settings for the control
of the tenant cache and throttling of the
identity-sdk requests.

koevskinikola committed:
feat: itegrate new config with interceptor classes

The new IdentityRequestCfg config properties were added
to enable the configuration of the tenant cache and
identity-sdk throttling of tenant-related requests. This
commit passes these config properties to the relevant
classes that perform the tenant-related requests.

koevskinikola committed:
test: implement the new identity-related properties

Pass along the new IdentityRequestCfg properties to the
relevant tests.

koevskinikola committed:
refactor: remove jetbrains notnull import

koevskinikola commented:
ℹ️ Implementation notes:
- I decided to implement the fix starting with `8.6` as with `8.7` Identity will be restructured and merged into Zeebe. So adding the fix into `8.7-SNAPSHOT` will create more clean-up work once we migrate to the new Identity architecture.
- I added an `ExperimentalCfg` to the gateway for the configuration properties used by the `tenant` cache and `semaphore`. This might be useful for any future work, especially now that we have a lot more teams using the Gateway.
- I could reproduce the issue with the IT test, but wasn't able to easily `assert` it. The `IOException` gets lost somewhere in the "gRPC translation" to the client. However, with debugging I could consistently reproduce the occurrence in the gRPC `IdentityInterceptor`. I'm open to any improvements to the code there.
- I named the `identity-sdk` wrapper class `IdentityTenantService`. It contains both a `LoadingCache` and a `semaphore`. I'm open to suggestions in regards to the naming and configuration defaults. I derived the defaults from the minimal amount of concurrent requests that triggered the bug in the above IT test.

npepinpe commented on zeebe/gateway/src/main/java/io/camunda/zeebe/gateway/impl/identity/IdentityTenantService.java:
```diff
@@ -0,0 +1,80 @@
+  private final Identity identity;
+
+  private final boolean isCachingEnabled;
+  private final long semaphoreTimeout;
```
:wrench: I would keep calling it by its logical name, `tenantRequestTimeout`.

npepinpe commented on zeebe/gateway/src/main/java/io/camunda/zeebe/gateway/impl/identity/IdentityTenantService.java:
```diff
@@ -0,0 +1,80 @@
+                  }
+                });
+    ree =
+        new RejectedExecutionException(
```
:question: So the stack trace will always show this right? We don't know where exactly it was thrown. I guess it's OK; not ideal, but it's definitely cheaper all things considered :thinking:

koevskinikola replied:
@npepinpe considering your comment [here](https://github.com/camunda/camunda/pull/24196#discussion_r1822944741), I'll only throw it if the semaphore lock can't be acquired (on line 67). Does that make sense?

koevskinikola replied:
UPDATE: I also created a new Zeebe `ConcurrentRequestException`, so it's something unique that can be mapped to this feature.

npepinpe commented on zeebe/gateway/src/main/java/io/camunda/zeebe/gateway/impl/identity/IdentityTenantService.java:
```diff
@@ -0,0 +1,80 @@
+                });
+    ree =
+        new RejectedExecutionException(
+            String.format(
```
:wrench: The message only explains what happens. I would update it to match our normal style, and also to indicate why this could have happened, and what a user can do, i.e. enable a self-service approach for a user. Something like:

```
Expected to fetch tenants from Identity within %d%s, but there are too many concurrent requests; either increase the tenant request capacity, or scale Identity to complete requests faster.
```

npepinpe commented on zeebe/gateway/src/main/java/io/camunda/zeebe/gateway/impl/identity/IdentityTenantService.java:
```diff
@@ -0,0 +1,80 @@
+      return getTenantsForTokenInternal(token);
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw ree;
```
:question: I'm not sure this is the right thing to do here. If the thread was interrupted, it doesn't mean that you timed out, it just means the caller thread is interrupted/likely to be shutdown. In most cases, I suspect throwing here doesn't really make a difference. 

:wrench: At any rate, I would recommend throwing a new exception, if only to keep the cause so we have a proper stack trace chain.

npepinpe replied:
Would it make sense to have a specific type thrown in this service for all cases where we couldn't grab the lease? Or would we handle them differently?

npepinpe commented on zeebe/gateway/src/main/java/io/camunda/zeebe/gateway/impl/identity/IdentityTenantService.java:
```diff
@@ -0,0 +1,80 @@
+      }
+      return getTenantsForTokenInternal(token);
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
```
:+1: :rocket:

npepinpe commented on zeebe/gateway/src/main/java/io/camunda/zeebe/gateway/impl/configuration/IdentityRequestCfg.java:
```diff
@@ -0,0 +1,97 @@
+
+import java.util.Objects;
+
+public class IdentityRequestCfg {
```
:upside_down_face: imo this is more of a `IdentityServiceCfg`, but honestly this is very much a nit :melting_face:

koevskinikola replied:
Sounds better to me, I'll refactor it.

remcowesterhoud commented on zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/interceptors/impl/IdentityInterceptor.java:
```diff
@@ -30,6 +32,7 @@ public final class IdentityInterceptor implements ServerInterceptor {
   private static final Metadata.Key<String> AUTH_KEY =
       Metadata.Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER);
   private final Identity identity;
```
💭 If we also put the token verification in the wrapper we don't need his field here anymore and it all goes through the wrapper. That way we don't differentiate between sdk calls, which could cause confusion.

koevskinikola replied:
The problem is that this code is gRPC-specific. My goal with the wrapper was to make it reusable for gRPC and the REST API. I would need to play around with lambdas (maybe even generics) to achieve that. 

It will probably be a nicer solution than what is currently provided, but we need to deliver a fix for this by next week, and I don't want to risk introducing a regression.

@remcowesterhoud would it be acceptable if I create a `kind/toil` issue for this?

remcowesterhoud replied:
Maybe I'm missing something, but I thought the `IdentityAuthenticationManager` was the rest part? They both make a call to `identity.authentication().verifyToken(tokenValue);` which seems like we could move it to the wrapper.

Anyway it was only a thought, feel free to ignore it.

koevskinikola replied:
Ah okay, I misunderstood. I thought you wanted to wrap the token extraction from the request as well.

During the solution discussion with @npepinpe, we decided to avoid adding other `identity-sdk` requests to this wrapper since no problems have been reported about them. We can expand the wrapper if additional issues are reported.

In the end, this fix is only a band-aid. The best solution would be to make Identity more scalable and we are currently making Identity more scalable by integrating its data model in Zeebe. After we're finished with the re-architecture, this fix won't be needed.

npepinpe commented on zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/interceptors/impl/IdentityInterceptor.java:
```diff
@@ -97,24 +113,32 @@ public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
-              .augmentDescription(
-                  "Expected Identity to provide authorized tenants, see cause for details")
-              .withCause(e));
+    } catch (final RejectedExecutionException ree) {
```
:wrench: I would propose using a specific exception type if only to avoid a possible `RejectedExcutionException` coming from the Identity SDK itself, which also uses some stuff internally via the HttpClient.

npepinpe commented on zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/multitenancy/MultiTenancyOverIdentityHighRequestCountIT.java:
```diff
@@ -0,0 +1,450 @@
+      final List<Future<ZeebeFuture<ProcessInstanceEvent>>> results =
+          executorService.invokeAll(piList);
+
+      Thread.sleep(10000);
```
:fearful: This looks like flakiness waiting to happen. Can we come up with a different approach?

koevskinikola replied:
I'll remove it, the `ExecutorService#invokeAll(...)` method actually blocks until all tasks are completed. I missed that.

npepinpe commented on zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/multitenancy/MultiTenancyOverIdentityHighRequestCountIT.java:
```diff
@@ -0,0 +1,450 @@
+            .done();
+  }
+
+  // Regression test: https://github.com/camunda/camunda/issues/23853
+  @Test
```
```suggestion
  @RegressionTest("https://github.com/camunda/camunda/issues/23853")
```

npepinpe commented on zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/multitenancy/MultiTenancyOverIdentityHighRequestCountIT.java:
:thought_balloon: A lot of the code here is reused from `MultiTenancyOverIdentityIT`. Normally I would propose to not duplicate, esp. things like the `PostgresHelper`, but I guess if we're getting rid of much of this later, maybe this is the right approach for now.

koevskinikola replied:
Yup, in the interest of time, I copied most of it. In turn, I think the `MultiTenancyOverIdentityIT` code is copied from `GatewayAuthenticationIdentityIT`.

Should I create a follow-up `kind/toil` issue for this cleanup?

koevskinikola replied:
FYI: https://github.com/camunda/camunda/issues/24254

remcowesterhoud commented on zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/interceptors/impl/IdentityTenantService.java:
```diff
@@ -0,0 +1,73 @@
+
+  private final int semaphoreCapacity = 300;
+  private final long semaphoreTimeout = 1;
+  private final long cacheSize = 1000;
```
❓ How did you decide on the size? It feels like 1000 is a lot, but then again maybe it isn't. I've got no clue 😅

koevskinikola replied:
That's a good point, I didn't have a good gauge on what would be a good value here. The cache size is directly correlated to the number of users, i.e. Identity JWT tokens, and that number may vary significantly per Zeebe user.

I agree, 1000 seems like a very large number for JWT tokens. I'll try to poll the Identity team for some metrics so I can derive some sensible number.

npepinpe replied:
Well, we can also see it from a memory point of view. Can we estimate how much 1000 entries would take? If it's still low, than that's fine I think.

koevskinikola replied:
I'm not sure it is. We're storing a list of tenants, so if there is a large number of tenants (let's say 200-300?) mapped to 1000 JWT tokens, is this still a low memory footprint?

koevskinikola replied:
Considering the average size of a JWT string (cache key) is ~1KB, and a single `tenantId` with max size (32 UTF-8 chars) is ~56bytes, plus the overhead of the Cache objects (let's say ~32 bytes if it's similar to a `Map.Entry`), a single entry in the cache would be ~1,1KB.

- Let's make an assumption that an installation has ~50 tenants (that's still quite a large number). If they are all assigned to a single JWT token, that's 50x56bytes=2,8KB.
- For 1000 entries where all JWT tokens are assigned to all tenants: 1000x(1,1KB+2,8KB+32bytes)=**3,9 MB**

I'll drop the default size to 200 entries. It's still a large number of JWT tokens, but the memory footprint should be ~780 KB then.

npepinpe reviewed (approved):
:rocket: Great stuff!

Generally I think we could probably unit test the `IdentityTenantService`, since the only new tests is very broad, and likely difficult to diagnose when things go wrong. Still pre-approving because all-in-all I don't think we need another review (but feel free to ask for one). Do address all comments before merging.

remcowesterhoud commented on zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/interceptors/impl/IdentityTenantService.java:
```diff
@@ -0,0 +1,73 @@
+  private final Semaphore semaphore;
+  private final Identity identity;
+
+  private final int semaphoreCapacity = 300;
```
❓ Again, how did you decide on this? Do we have some numbers from which we found a sensible limit?

koevskinikola replied:
In the IT test, the bug wasn't triggered if less than 300 concurrent requests were sent to the gateway, and between 300-500 the test/bug occurrence became flaky. So I think 300 is a nice limit to prevent entering the "flaky zone".

remcowesterhoud commented on dist/src/main/java/io/camunda/zeebe/shared/security/IdentityAuthenticationManager.java:
```diff
@@ -25,13 +27,15 @@
 public final class IdentityAuthenticationManager implements AuthenticationManager {
 
   private final Identity identity;
```
💭 Same comment here regarding wrapping all the identity sdk call into the service

koevskinikola replied:
See my answer [here](https://github.com/camunda/camunda/pull/24196#discussion_r1824102403).

remcowesterhoud commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RestErrorMapper.java:
```diff
@@ -142,18 +143,22 @@ yield createProblemDetail(
       case final PartitionInactiveException pie:
         final var pieMsg =
-            "Expected to handle gRPC request, but the target partition is currently inactive";
+            "Expected to handle REST API request, but the target partition is currently inactive";
```
👍 Interesting that this wasn't caught before 😅

remcowesterhoud reviewed (approved):
Thanks @koevskinikola, no blockers from my side but a few comments 👀

koevskinikola committed:
refactor: apply review feedback for identity-sdk wrapper

- Rename configuration class and various properties,
- Use Zeebe-specific exceptions,
- Adjust tests.

koevskinikola committed:
refactor: implement review feedback for identity-sdk wrapper

- Reduce tenant cache size from 1000 to 200.
  A cache of 200 JWT tokens is still large, but doesn't
  have a large memory footprint.
- Increase the number of concurrent requests in the IT
  tests to trigger the concurrent requests issue while
  avoiding test flakiness.

koevskinikola commented on zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/multitenancy/MultiTenancyOverIdentityHighRequestCountIT.java:
```diff
@@ -0,0 +1,449 @@
+                    }
+                  })
+              .map(ZeebeFuture::toCompletableFuture)
+              .filter(CompletableFuture::isCompletedExceptionally)
+              .map(CompletableFuture::exceptionNow)
```
@remcowesterhoud I tried adding an additional assertion out, but it seems like a high load is not always reached in Zeebe. I think that's acceptable, as the issue is in Identity not being able to handle high load, and the wrapper is handling that.

koevskinikola committed:
refactor: restore it test assertions for identity-sdk wrapper

Only assert that no Identity-related exceptions occured.
It is acceptable if no Zeebe-related exceptionds due to
high load occurred.

koevskinikola committed:
docs: document the configuration options for the identity wrapper

Add documentation on the experimental identity wrapper
configuration properties in out YAML templates.

backport-action commented:
Created backport PR for `stable/8.4`:
- #24269 with remaining conflicts!

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin backport-24196-to-stable/8.4
git worktree add --checkout .worktree/backport-24196-to-stable/8.4 backport-24196-to-stable/8.4
cd .worktree/backport-24196-to-stable/8.4
git reset --hard HEAD^
git cherry-pick -x c429fa3ddb1a6214fb4a467c24b21b78a545775f 7183d92d9eed17842919cbdb2cd164bdc2e16c91 57e34e6f031fa97b15c5a61b6c89028ce145cbb8 5db948564c72eff6210e49236de4410737b28a6a 9efd8cce2bc5ef4b894a25cab1176e45c6438f3a c4784bb6e0dfc5421f4f274392a6bcd3500aba1b 7dce0ca2fc71a3ce952d895ad56baf2d19ebeecc 27b77265d95011b44fcbffbd23ddbaab9c3c2838 2c1182be74d6382265749f7b5376b9f5f4b25413 80966b1eb75942f075729592e8e309f9d7715956 088a826b8727ad557a37ab12de273a716348f6e8 c2fc40716f93cd33066e5fb61d094284d80386d0
git push --force-with-lease
```

backport-action commented:
Created backport PR for `stable/8.5`:
- #24270 with remaining conflicts!

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin backport-24196-to-stable/8.5
git worktree add --checkout .worktree/backport-24196-to-stable/8.5 backport-24196-to-stable/8.5
cd .worktree/backport-24196-to-stable/8.5
git reset --hard HEAD^
git cherry-pick -x 7183d92d9eed17842919cbdb2cd164bdc2e16c91 57e34e6f031fa97b15c5a61b6c89028ce145cbb8 5db948564c72eff6210e49236de4410737b28a6a 9efd8cce2bc5ef4b894a25cab1176e45c6438f3a c4784bb6e0dfc5421f4f274392a6bcd3500aba1b 7dce0ca2fc71a3ce952d895ad56baf2d19ebeecc 27b77265d95011b44fcbffbd23ddbaab9c3c2838 2c1182be74d6382265749f7b5376b9f5f4b25413 80966b1eb75942f075729592e8e309f9d7715956 088a826b8727ad557a37ab12de273a716348f6e8 c2fc40716f93cd33066e5fb61d094284d80386d0
git push --force-with-lease
```
